core/storage/chain/impl/chain_store_impl.cpp

```diff
@@ -3,56 +3,55 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-#include "storage/chain/chain_store.hpp"
+#include "storage/chain/impl/chain_store_impl.hpp"
 
-#include "common/le_encoder.hpp"
 #include "common/outcome.hpp"
-#include "datastore_key.hpp"
+#include "crypto/randomness/impl/chain_randomness_provider_impl.hpp"
+#include "storage/chain/datastore_key.hpp"
 #include "primitives/address/address_codec.hpp"
 #include "primitives/cid/cid_of_cbor.hpp"
 #include "primitives/cid/json_codec.hpp"
-#include "primitives/persistent_block/persistent_block.hpp"
 #include "primitives/tipset/tipset_key.hpp"
 
 namespace fc::storage::blockchain {
+  using crypto::randomness::ChainRandomnessProviderImpl;
   using primitives::block::BlockHeader;
-  using primitives::blockchain::block::PersistentBlock;
   using primitives::tipset::Tipset;
 
   namespace {
     const DatastoreKey chain_head_key{DatastoreKey::makeFromString("head")};
     const DatastoreKey genesis_key{DatastoreKey::makeFromString("0")};
   }  // namespace
 
-  ChainStore::ChainStore(std::shared_ptr<ipfs::IpfsBlockService> block_service,
-                         std::shared_ptr<ChainDataStore> data_store,
-                         std::shared_ptr<BlockValidator> block_validator,
-                         std::shared_ptr<WeightCalculator> weight_calculator)
+  ChainStoreImpl::ChainStoreImpl(std::shared_ptr<ipfs::IpfsBlockService> block_service,
+                                 std::shared_ptr<ChainDataStore> data_store,
+                                 std::shared_ptr<BlockValidator> block_validator,
+                                 std::shared_ptr<WeightCalculator> weight_calculator)
       : block_service_{std::move(block_service)},
         data_store_{std::move(data_store)},
         block_validator_{std::move(block_validator)},
         weight_calculator_{std::move(weight_calculator)} {
     logger_ = common::createLogger("chain store");
   }
 
-  outcome::result<std::shared_ptr<ChainStore>> ChainStore::create(
+  outcome::result<std::shared_ptr<ChainStoreImpl>> ChainStoreImpl::create(
       std::shared_ptr<ipfs::IpfsBlockService> block_store,
       std::shared_ptr<ChainDataStore> data_store,
       std::shared_ptr<BlockValidator> block_validator,
       std::shared_ptr<WeightCalculator> weight_calculator) {
-    std::shared_ptr<ChainStore> cs{};
-    ChainStore tmp(std::move(block_store),
-                   std::move(data_store),
-                   std::move(block_validator),
-                   std::move(weight_calculator));
-    cs.reset(new ChainStore(std::move(tmp)));
+    std::shared_ptr<ChainStoreImpl> cs{};
+    ChainStoreImpl tmp(std::move(block_store),
+                       std::move(data_store),
+                       std::move(block_validator),
+                       std::move(weight_calculator));
+    cs.reset(new ChainStoreImpl(std::move(tmp)));
```

Comment:
```
std::shared_ptr<ChainStoreImpl> cs =
        std::make_shared<ChainStoreImpl>(std::move(block_store),
                                         std::move(data_store),
                                         std::move(block_validator),
                                         std::move(weight_calculator));
```

Reply:
This is not that simple. ChainStoreImpl has private constructor, which makes too complicated creating via std::make_shared. However, I refactored it as much as it was possible.